### PR TITLE
Read minimum staking value from database

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- The snapshot can be made on the past blocks as well. The minimum staking parameter can be different on the past block, but the minimum staking parameter is stored and used globally. This PR fixes this change.
- This PR should be merged after #1018.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: governance/handler.go [hunk 1]
```diff
@@ -459,3 +459,12 @@ func (gov *Governance) GetGoverningInfoAtNumber(num uint64) (bool, common.Addres
...
 	return true, govNode.(common.Address), nil
 }
+
+func (gov *Governance) GetMinimumStakingAtNumber(num uint64) (uint64, error) {
+	minStaking, err := gov.GetItemAtNumberByIntKey(num, params.MinimumStake)
+	if err != nil {
+		return 0, err
+	}
+	bigMinStaking, _ := new(big.Int).SetString(minStaking.(string), 10)
+	return bigMinStaking.Uint64(), nil
```
Comment (hqjang-pepper) [inline governance/handler.go:469]:
```suggestion
	if bigMinStaking, ok := new(big.Int).SetString(minStaking.(string), 10); ok {
		return bigMinStaking.Uint64(), nil	
	}
```

How about checking if func SetString returns true?

Comment (jeongkyun-oh) [reply governance/handler.go]: Reflected. I have added an error case.
Thanks :)

File: consensus/istanbul/backend/engine_test.go [hunk 2]
```diff
@@ -644,6 +653,126 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+		validators []int
+		demoted    []int
+	}
+	type testcase struct {
+		stakingInfo []uint64
+		votes       []vote
+		expected    []expected
+	}
+
+	testcases := []testcase{
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine_test.go:672]: @jeongkyun-oh Could you please check below tests cases?

Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go:672]: @ehnuje The test cases are enhanced and updated. Please take another look.

Comment (ehnuje) [reply consensus/istanbul/backend/engine_test.go:672]: Thanks! It would be appreciated if the purpose of each test case is described.

**Review by ehnuje — DISMISSED** [review]

**Review by hqjang-pepper — APPROVED** [review]

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -644,6 +653,156 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+			[]vote{
+				{"governance.governancemode", "none"}, // voted on epoch 1, applied from 6-8
+				{"reward.minimumstake", "5500000"},    // voted on epoch 2, applied from 9-11
+				{"reward.minimumstake", "6500000"},    // voted on epoch 3, applied from 12-14
+				{"reward.minimumstake", "7500000"},    // voted on epoch 4, applied from 15-17
+				{"reward.minimumstake", "8500000"},    // voted on epoch 5, applied from 18-20
+				{"reward.minimumstake", "7500000"},    // voted on epoch 6, applied from 21-23
+				{"reward.minimumstake", "6500000"},    // voted on epoch 7, applied from 24-26
+				{"reward.minimumstake", "5500000"},    // voted on epoch 8, applied from 27-29
+				{"reward.minimumstake", "4500000"},    // voted on epoch 8, applied from 30-32
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: I have updated. Please take anoter look.

File: consensus/istanbul/backend/engine_test.go [hunk 4]
```diff
@@ -644,6 +653,156 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+			[]vote{
+				{"reward.minimumstake", "8500000"}, // voted on epoch 1, applied from 6-8
+				{"reward.minimumstake", "7500000"}, // voted on epoch 2, applied from 9-11
+				{"reward.minimumstake", "6500000"}, // voted on epoch 3, applied from 12-14
+				{"reward.minimumstake", "5500000"}, // voted on epoch 4, applied from 15-17
+				{"reward.minimumstake", "4500000"}, // voted on epoch 5, applied from 18-20
+				{"reward.minimumstake", "5500000"}, // voted on epoch 6, applied from 21-23
+				{"reward.minimumstake", "6500000"}, // voted on epoch 7, applied from 24-26
+				{"reward.minimumstake", "7500000"}, // voted on epoch 8, applied from 27-29
+				{"reward.minimumstake", "8500000"}, // voted on epoch 8, applied from 30-32
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: I have updated. Please take anoter look.

File: governance/handler.go [hunk 5]
```diff
@@ -54,7 +55,7 @@ var GovernanceItems = map[int]check{
 	params.Ratio:                   {stringT, checkRatio, nil},
 	params.UseGiniCoeff:            {boolT, checkUint64andBool, updateUseGiniCoeff},
 	params.DeferredTxFee:           {boolT, checkUint64andBool, nil},
-	params.MinimumStake:            {stringT, checkBigInt, updateMinimumStakingAmount},
+	params.MinimumStake:            {stringT, checkBigInt, nil},
```
Comment (aidan-kwon) [inline governance/handler.go:58]: Why did you remove updateMinimumStakingAmount? I think this change makes the governance item unavailable.

Comment (jeongkyun-oh) [reply governance/handler.go:58]: The `updateMinimumStakingAmount` is used to update global minimum staking variable which is not used anymore. We get the minmum staking value from database, so I removed this trigger. FYI, the current set still exists and is updated properly on `UpdateCurrentSet` function.

File: consensus/istanbul/backend/engine_test.go [hunk 6]
```diff
@@ -80,7 +80,8 @@ func makeCommittedSeals(hash common.Hash) [][]byte {
 func newBlockChain(n int, items ...interface{}) (*blockchain.BlockChain, *backend) {
 	// generate a genesis block
 	genesis := blockchain.DefaultGenesisBlock()
-	genesis.Config = params.TestChainConfig
+	config := *params.TestChainConfig // copy test chain config which may be modified
+	genesis.Config = &config
```
Comment (hqjang-pepper) [inline consensus/istanbul/backend/engine_test.go:84]: It seems good to import chainconfig like this way in case the chainconfig changes later. LGTM

**Review by hqjang-pepper — DISMISSED** [review]

File: consensus/istanbul/backend/engine_test.go [hunk 7]
```diff
@@ -654,6 +655,159 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+				{"reward.minimumstake", "6500000"}, // voted on epoch 3, applied from 12-14
+				{"reward.minimumstake", "5500000"}, // voted on epoch 4, applied from 15-17
+				{"reward.minimumstake", "4500000"}, // voted on epoch 5, applied from 18-20
+				{"reward.minimumstake", "5500000"}, // voted on epoch 6, applied from 21-23
+				{"reward.minimumstake", "6500000"}, // voted on epoch 7, applied from 24-26
+				{"reward.minimumstake", "7500000"}, // voted on epoch 8, applied from 27-29
+				{"reward.minimumstake", "8500000"}, // voted on epoch 9, applied from 30-32
+			},
+			[]expected{
+				// 0 is governing node, so it is included in th validators all the time
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:713]:
```suggestion
				// 0 is governing node, so it is included in the validators all the time
```

File: consensus/istanbul/backend/engine_test.go [hunk 8]
```diff
@@ -654,6 +655,159 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+				{[]uint64{9, 10, 11}, []int{0, 3}, []int{1, 2}},
+				{[]uint64{12, 13, 14}, []int{0, 2, 3}, []int{1}},
+				{[]uint64{15, 16, 17, 18, 19, 20, 21, 22, 23}, []int{0, 1, 2, 3}, []int{}},
+				{[]uint64{24, 25, 26}, []int{0, 2, 3}, []int{1}},
+				{[]uint64{27, 28, 29}, []int{0, 3}, []int{1, 2}},
+				{[]uint64{30, 31, 32}, []int{0, 1, 2, 3}, []int{}},
+			},
+		},
+		{
+			// test the validators are updated properly if governing node is changed
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:724]:
```suggestion
		{ // test the validators are updated properly if governing node is changed	
```

File: consensus/istanbul/backend/engine_test.go [hunk 9]
```diff
@@ -654,6 +655,159 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+			[]uint64{6000000, 6000000, 5000000, 5000000},
+			[]vote{
+				{"reward.minimumstake", "5500000"}, // voted on epoch 1, applied from 6-8
+				{"governance.governingnode", 2},    // voted on epoch 2, applied from 9-11
+			},
+			[]expected{
+				// 0 is governing node, so it is included in th validators all the time
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:731]:
```suggestion
				// 0 is governing node, so it is included in the validators all the time
```

File: consensus/istanbul/backend/engine_test.go [hunk 10]
```diff
@@ -654,6 +655,159 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
...
+				assert.Equal(t, expectedDemoted, demoted)
+			}
+		}
+
+		reward.SetTestStakingManager(oldStakingManager)
+		engine.Stop()
+	}
+}
+
+func TestSnapshot_MinimumStaking2(t *testing.T) {
```
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:810]: With the name of `minimumStaking2`, it is hard to figure out what this test do. Could you change the name of `TestSnapshot_MinimumStaking2`?

File: governance/handler.go [hunk 11]
```diff
@@ -462,3 +463,15 @@ func (gov *Governance) GetGoverningInfoAtNumber(num uint64) (bool, common.Addres
 
 	return true, govNode.(common.Address), nil
 }
+
+func (gov *Governance) GetMinimumStakingAtNumber(num uint64) (uint64, error) {
+	minStaking, err := gov.GetItemAtNumberByIntKey(num, params.MinimumStake)
```
Comment (hyochan-brown) [inline governance/handler.go:468]: Previously, the naming minimumStaking was used. 
But the naming minStaking appeared in this PR. 
Does it have a different meaning?

File: consensus/istanbul/backend/engine_test.go (hunk 8 above)
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:724]:
```suggestion
		{ // test the validators are updated properly if governing node is changed
```

File: consensus/istanbul/backend/engine_test.go (hunk 7 above)
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:713]:
```suggestion
				// 0 is governing node, so it is included in the validators all the time
```

File: consensus/istanbul/backend/engine_test.go (hunk 9 above)
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:731]:
```suggestion
				// 0 is governing node, so it is included in the validators all the time
```

File: consensus/istanbul/backend/engine_test.go (hunk 10 above)
Comment (yoomee1313) [inline consensus/istanbul/backend/engine_test.go:810]: With the name of `minimumStaking2`, it is hard to figure out what this test do. Could you change the name of
`TestSnapshot_MinimumStaking` and `TestSnapshot_MinimumStaking2`?

**Review by yoomee1313 — DISMISSED** [review]
@jeongkyun-oh ltgm except typo

File: governance/handler.go (hunk 11 above)
Comment (jeongkyun-oh) [reply governance/handler.go:469]: Nope, it is the same. Dosen't it clear `min` is an abbreviation of `minimum`?

File: consensus/istanbul/backend/engine_test.go (hunk 8 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go:726]: I made the comments consistent with others. PTAL.

File: consensus/istanbul/backend/engine_test.go (hunk 7 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: Updated. Thanks.

File: consensus/istanbul/backend/engine_test.go (hunk 9 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: Updated. Thanks.

File: consensus/istanbul/backend/engine_test.go (hunk 10 above)
Comment (jeongkyun-oh) [reply consensus/istanbul/backend/engine_test.go]: Updated the test names. Thanks.

File: governance/handler.go (hunk 11 above)
Comment (hyochan-brown) [reply governance/handler.go:469]: I think it's clear.
I just asked a question because it was different from before.
Thanks!!

**Review by yoomee1313 — APPROVED** [review]

**Review by hqjang-pepper — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]
